gordon/plugins_loader.py

```diff
@@ -186,6 +186,14 @@ def _gather_installed_plugins():
     return gathered_plugins
 
 
+def _get_metrics_plugin(config, installed_plugins):
+    metrics_config = config.get('metrics', {})
+    for plugin_name, plugin in installed_plugins.items():
+        if plugin_name == metrics_config.get('provider', {}):
```

Comment:
Hm, this forces the service operator to be aware of the entry point name that the plugin itself has declared.

Could we have the config key/section be the name without the `gordon.metrics` prefix, i.g. `[ffwd]`? Then maybe add another optional `[core]` variable, e.g.

```toml
[core]
plugins = ["gcp.publisher", "..."]
# or metrics_provider, or whatever
metrics = "ffwd"

[ffwd]
host = "10.99.0.1"
port = 19000
```

so then the code above is similar:

```py
metrics_provider = config.get('metrics')

# maybe this part is at a later commit/PR?
if not metrics_provider:  
    # instantiate & return metrics logger
if metrics_provider == 'ffwd'
    # import ffwd, instantiate w/ config, & return

for plugin_name, plugin in installed_plugins.items():
   if plugin_name == metrics_provider:
        metrics_config = config.get(metrics_provider, {})
        # load, instantiate, return
```

This allows configuration to be abstracted away a bit better:

```toml
[core]
metrics = "prometheus"

[prometheus]
exporter = "http"  # or twisted, wsgi, pushgateway
multiprocess = true
authentication = "yesplease"
```

And plugin developers can still register their entry point w/ a simpler name:

```py
# prometheus_plugin/setup.py
setup(
    # snip
    entry_points={
        'gordon.plugins': [
            'prometheus = path.to:callable',
            # would otherwise be this w/ current setup
            'gordon.metrics.prometheus = path.to.callable'
        ],
    },
    # snip
)
```

Reply:
I like this.  It's also more consistent with the config style for the other plugins.

Reply:
@econchick I’m copying your comments here, but so I can address each one separately, and I've numbered them.  I'm hoping that will make it easier to discuss. My response to each comment is below the line.

>1) Hm, this forces the service operator to be aware of the entry point name that the plugin itself has declared.
---
Won't the service operator have to be aware of the entry point names for all of the plugins (since [they have to have config based on that name to be inited](https://github.com/spotify/gordon/pull/22/commits/c0b961fa349b34ff5ac361f799677b13820b7bbc#diff-41273d1d026664b64fbaab5565b7a9b5L78))?  I'm didn’t see anything that made metrics in particular different from the other plugins in this sense, so I followed the same pattern.  

> 2) Could we have the config key/section be the name without the gordon.metrics prefix, i.g. [ffwd]?
---
I don’t think there is a `gordon.metrics` section.  It should just be `metrics`, right?  I think this confusion is because we maybe picked a bad example name and used it in the naming in the tests, but there shouldn't be anything special about it, its just a string. I’ve updated it in the latest changes. Does it make more sense now?  

> 3) Then maybe add another optional [core] variable, e.g. …
```toml
[core]
plugins = ["gcp.publisher", "..."]
# or metrics_provider, or whatever
metrics = "ffwd"

[ffwd]
host = "10.99.0.1"
port = 19000
```
----
We did consider having a core variable for metrics provider name, and then a separate section for metrics configuration.  Then we decided there was no harm in passing name to the plugin, and it was nice to keep all the metrics configuration together. I think calling the section ‘metrics’, makes it more obvious what it’s for, but I don’t have too strong of an opinion and could go with ffwd/prometheus etc.  Just right now I'm sure why it would be better.

> 4) so then the code above is similar:
```python
metrics_provider = config.get('metrics')

# maybe this part is at a later commit/PR?
if not metrics_provider:  
    # instantiate & return metrics logger
if metrics_provider == 'ffwd':
    # import ffwd, instantiate w/ config, & return

for plugin_name, plugin in installed_plugins.items():
   if plugin_name == metrics_provider:
        metrics_config = config.get(metrics_provider, {})
        # load, instantiate, return
```
---
I think the first line is wrong, and I don’t really like special handling for ffwd, but overall I think it’s okay, but should be put off to a later PR when we have a default.  With these two changes I imagine it would be more like:
```
metrics_provider = config.get('core').get('metrics')
if not metrics_provider:  
    # instantiate & return metrics logger]

for plugin_name, plugin in installed_plugins.items():
   if plugin_name == metrics_provider:
        metrics_config = config.get(metrics_provider, {})
        # load, instantiate, return
```
 
> 5) This allows configuration to be abstracted away a bit better: 
---
This is depending on what we do in 3. I think your suggestion is a fine way to do it, but I’m not sure I agree the abstraction is better.  Although I like the consistency with the other plugins, I think that consistency could be confusing.  Because a metrics plugin is handled differently, it makes a little sense to configure it differently. In your example its less obvious you’re configuring something for metrics here, and could be more confused with the other plugins.  To suggest a third option, how do you feel about something like the following

```
[metrics.ffwd]
host = "10.99.0.1"
port = 19000
unit = 10E9
```
It allows it to be abstracted, but also namespaced for metrics? It would require only minimal code changes I think.

> 6) And plugin developers can still register their entry point w/ a simpler name:
---
As mentioned in 2, I think the name is not quite right. I believe it currently would be as you’re suggesting, i.e. the following.
```
# prometheus_plugin/setup.py
setup(
    # snip
    entry_points={
        'gordon.plugins': [
            ‘ffwd = path.to.callable'
        ],
    },
    # snip
)

Reply:
I'd like to talk about this in person since I'm not exactly following this (I'll be sure to update this comment here w/ what was discussed + the outcome).